python_scripts/ensemble_bagging.py

```diff
@@ -184,72 +181,156 @@ def bootstrap_sample(data, target):
+    base_estimator=DecisionTreeRegressor(max_depth=3),
+    n_estimators=100,
+)
+_ = bagged_trees.fit(data_train, target_train)
 
+# %% [markdown]
+#
+# Let us visualize the predictions of the ensemble on the same test data:
 # %%
 sns.scatterplot(x=data_train["Feature"], y=target_train, color="black",
                 alpha=0.5)
-plt.plot(data_test, target_predicted_forest, label="Bag of decision trees")
-plt.legend()
+plt.plot(data_test, bagged_trees.predict(data_test))
 _ = plt.title("Predictions from a bagging classifier")
 
 # %% [markdown]
-# While we used a decision tree as a base model, nothing prevent us of using
-# any other type of model. We will give an example that will use a linear
-# regression.
+#
+# Because we use 100 trees in the ensemble, the average prediction is indeed
+# slightly smoother but very similar to our previous average plot.
+#
+# It is possible to access the internel models of the ensemble in stored as a
+# Python list in the `bagged_trees.estimators_` attribute after fitting.
+#
+# Let us compare the based model predictions with their average:
 
 # %%
-from sklearn.linear_model import LinearRegression
+for tree_idx, tree in enumerate(bagged_trees.estimators_):
+    label = "Predictions of individual trees" if tree_idx == 0 else None
+    plt.plot(data_test, tree.predict(data_test), linestyle="--", alpha=0.1,
+             color="tab:blue", label=label)
+
+sns.scatterplot(x=data_train["Feature"], y=target_train, color="black",
+                alpha=0.5)
+
+plt.plot(data_test, bagged_trees.predict(data_test),
+         color="tab:orange", label="Predictions of ensemble")
+_ = plt.legend()
 
-bagging = BaggingRegressor(base_estimator=LinearRegression(),
-                           n_estimators=3)
-bagging.fit(data_train, target_train)
-target_predicted_linear = bagging.predict(data_test)
+# %% [markdown]
+#
+# We used a low value of the opacity parameter `alpha` to better appreciate the
+# overlap in the prediction functions of the individual trees.
+#
+# This visualization gives some insights on the uncertainty in the predictions
+# in different areas of the feature space.
+#
+# ## Bagging complex pipelines
+#
+# While we used a decision tree as a base model, nothing prevents us of using
+# any other type of model.
+#
+# As we know that the original data generating function is a noisy polynomial
+# transformation of the input variable, let us try to fit a bagged polynomial
+# regression pipeline on this dataset:
 
 # %%
+from sklearn.linear_model import Ridge
+from sklearn.preprocessing import PolynomialFeatures
+from sklearn.preprocessing import MinMaxScaler
+from sklearn.pipeline import make_pipeline
+
+
+polynomial_regressor = make_pipeline(
+    MinMaxScaler(),
+    PolynomialFeatures(degree=4),
+    Ridge(alpha=1e-10),
+)
+bagging = BaggingRegressor(
+    base_estimator=polynomial_regressor,
+    n_estimators=100,
+    random_state=0,
+)
+_ = bagging.fit(data_train, target_train)
```

Comment:
```suggestion
polynomial_regressor = make_pipeline(
    MinMaxScaler(),
    PolynomialFeatures(degree=4),
    Ridge(alpha=1e-10),
)

# %% [markdown]
# The polynomial regression pipeline will first scale each feature between 0 and 1
# and then expend the feature space by computing polynomial and interaction
# features up to a fourth degree. Now, we can use this model in a bagging ensemble:

# %%
bagging = BaggingRegressor(
    base_estimator=polynomial_regressor,
    n_estimators=100,
    random_state=0,
)
_ = bagging.fit(data_train, target_train)
```

Reply:
There are no interaction features in this case (single feature dataset).

Reply:
I will push an alternative.